src/mds/MDSDmclockScheduler.cc

```diff
@@ -0,0 +1,891 @@
+
+    std::unique_lock<std::mutex> lock(queue_mutex);
+
+    request_queue.emplace_back(new UpdateRequest(vid, client_info, use_default, cb_func));
+    /* wake up*/
+    lock.unlock();
+    queue_cvar.notify_all();
+}
+
+void MDSDmclockScheduler::process_request_handler()
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);
+
+  if (request_queue.size() == 0) {
+    queue_cvar.wait(lock);
+  }
+
+  while (request_queue.size()) {
+    std::unique_ptr<Request> request = std::move(request_queue.front());
+    request_queue.erase(request_queue.begin());
+
+    dout(10) << __func__ << " Process request type " << request->get_request_type_str() << dendl;
+
+    lock.unlock();
+
+    switch(request->get_request_type()) {
+      case RequestType::CLIENT_REQUEST:
+      {
+        std::unique_ptr<ClientRequest> c_request(static_cast<ClientRequest *>(request.release()));
+
+        dout(10) << " Process client request (queue size " << request_queue.size()
+                << " volume_id " << c_request->get_volume_id()
+                << " time " << c_request->time
+                << " cost " << c_request->cost << ")" << dendl;
+
+        increase_inflight_request(c_request->get_volume_id());
+
+        dmclock_queue->add_request(std::move(c_request), std::move(c_request->get_volume_id()),
+            {0, 0}, c_request->time, c_request->cost);
+
+        break;
+      }
+      case RequestType::UPDATE_REQUEST:
+      {
+        std::unique_ptr<UpdateRequest> c_request(static_cast<UpdateRequest *>(request.release()));
+
+        dout(10) << " Process update request (queue size " << request_queue.size()
+                  << " volume_id " << c_request->get_volume_id() << ")" << dendl;
+
+        update_volume_info(c_request->get_volume_id(), c_request->client_info, c_request->use_default);
+        dmclock_queue->update_client_info(c_request->get_volume_id());
+        set_volume_info_updated(c_request->get_volume_id());
+
+        if (c_request->cb_func) {
+          c_request->cb_func();
+        }
+        break;
+      }
+    }
+
+    lock.lock();
+  }
+}
+
+void MDSDmclockScheduler::process_request()
+{
+  dout(10) << __func__ << " thread has been invoked" << dendl;
+  while (state == SchedulerState::RUNNING) {
+    process_request_handler();
+  }
+  dout(10) << __func__ << " thread has been joined" << dendl;
+}
+
+void MDSDmclockScheduler::begin_schedule_thread()
+{
+  scheduler_thread = std::thread([this](){process_request();});
+}
+
+void MDSDmclockScheduler::enable_qos_feature()
+{
+  dout(10) << __func__ << dendl;
+
+  default_conf.set_status(true);
+
+  if (!get_mds_is_active()) {
+    dout(0) << " MDS Rank is not active and QoS feature cannot be enabled." << dendl;
+    return;
+  }
+
+  auto sessionmap = get_session_map();
+  if (sessionmap == nullptr) {
+    dout(10) << __func__ << " sessionmap has nullptr" << dendl;
+    return;
+  }
+
+  if (auto it = sessionmap->by_state.find(Session::STATE_OPEN); it != sessionmap->by_state.end()) {
```

Comment:
does this traverse the sessionmap twice? is there a more efficient way to traverse it once for OPEN | STALE?
(thinking of recent work with many thousands of clients.

Reply:
Each state maintains its own list. It seems difficult to iterate at once.
`std::map<int,xlist<Session*>*> by_state`
https://github.com/ceph/ceph/blob/main/src/mds/SessionMap.h#L805
https://github.com/ceph/ceph/blob/main/src/mds/SessionMap.cc#L1062

It seems that it can be simplified a little by making a { OPEN, STALE } list and using a double iteration. Any other comments?